tests/integration/page-router.test.ts

```diff
@@ -153,3 +153,124 @@ test<FixtureTestContext>('Should serve correct locale-aware custom 404 pages', a
     'Served 404 page content should use non-default locale if non-default locale is explicitly used in pathname (after basePath)',
   ).toBe('fr')
 })
+
+describe.only('404 caching', () => {
+  describe('404 without getStaticProps', () => {
```

Comment:
I found it was necessary to keep these tests because handling is different for 404 without `getStaticProps`, however I did remove cases for 'matching dynamic path without revalidate' because behaviour should always be the same as 'matching dynamic path with revalidate' and cache handler should be called in both cases